bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,409 @@
+    _all_runqueues: weakref.WeakSet = weakref.WeakSet()
+    _all_workers: weakref.WeakSet = weakref.WeakSet()
+
+    @classmethod
+    def _on_interpreter_shutdown(cls) -> None:
+        # Gracefully shutdown daemonized threads
+        # upon interpreter shutdown.
+        with cls._lock:
+            cls._interpreter_shutdown = True
+            for q in cls._all_runqueues:
+                q.put(None)
+            cls._all_runqueues.clear()
+            for worker in cls._all_workers:
+                if worker.is_alive():
+                    worker.join()
+            cls._all_workers.clear()
+
+    @dataclasses.dataclass
+    class Work:
+        future: concurrent.futures.Future
+        fn: typing.Callable[..., typing.Any]
+        args: typing.Tuple[typing.Any, ...]
+        kwargs: typing.Dict[str, typing.Any]
+
+        def execute(self) -> None:
+            try:
+                self.future.set_result(
+                    self.fn(*self.args, **self.kwargs))
+            except Exception as e:
+                self.future.set_exception(e)
+
+        def cancel(self) -> None:
+            self.future.cancel()
+
+        def cancelled(self) -> bool:
+            return self.future.cancelled()
+
+        def when_done(
+            self,
+            callback: typing.Callable[['AutoScalingThreadPool.Work'], None]
+        ) -> None:
+            self.future.add_done_callback(lambda f: callback(self))
+
+    def __init__(
+        self, *,
+        min_workers: typing.Optional[int] = None,
+        max_workers: typing.Optional[int] = None,
+        submission_quota: typing.Optional[int] = None,
+        submission_patience: typing.Optional[float] = None,
+        max_idle_time: typing.Optional[float] = None
+    ):
+        """
+        Initializes the thread pool.
+
+        :param min_workers: minimum number of workers in the pool, 0 by default.
+        :param max_workers: maximum number of workers in the pool, 32 times the
+          number of available CPU threads by default (assuming I/O bound work).
+        :param submission_quota: maximum number of concurrent submissions for a
+          a given callable. Up to the maximum number of workers by default. Useful
+          when serving multiple users to prevent anyone from starving the rest.
+        :param submission_patience: time to wait in seconds for a worker to become
+          available before upscaling the pool. 100 ms by default.
+        :param max_idle_time: time in seconds for a worker to wait for work before
+          shutting itself down, effectively downscaling the pool. 60 seconds by default.
+        """
+
+        if min_workers is None:
+            min_workers = 0
+        if min_workers < 0:
+            raise ValueError(
+                'Minimum number of workers '
+                'must be a nonnegative number')
+        self._min_workers = min_workers
+
+        if max_workers is None:
+            max_workers = 32 * (os.cpu_count() or 1)
```

Comment:
@jbarry-bdai on a typical Intel CPU with 4 physical cores, 8 logical threads, this amounts to 512 concurrent threads. That should be enough for most use cases. We can lift the restriction entirely too, but that many concurrent threads would suggest other issues.

Reply:
Haha yes this seems good.  The point is to make it easy for the general/casual user.

Reply:
> what does tuning actually look like for this for say an embedded low powered platform? Is there a way to stress test this end to end through ROS?

That is a great question I don't have a good answer for. One would think that the size of the pool would normally depend on latency requirements considering incoming data rates and processing rates, constrained by the resources you have. Blocking calls across callbacks adds the maximum depth of the associated call graphs to the set of affecting factors. So I'd probably start with these defaults, and guess my way out of trouble if it ever comes to that. You _could_ do a more thorough analysis by heavily tracing your system, and [ROS 2 has tracing support](https://github.com/ros2/ros2_tracing) for it. Not trivial though.

And then you have the _other_ parameters in this executor. Quotas are really just a way to mitigate a sudden imbalance between data rates and processing rates for reentrant callback groups (ie. for some reason you don't process things fast enough and the number of callbacks starts to pile up). You may not have any, which is why they are disabled by default (by making them equal to the maximum pool size). Patience is a bound on OS overhead for firing up threads, so OS dependent.